src/endOfYear/index.ts

```diff
@@ -15,20 +19,35 @@ import requiredArgs from '../_lib/requiredArgs/index'
  * - [Changes that are common for the whole library](https://github.com/date-fns/date-fns/blob/master/docs/upgradeGuide.md#Common-Changes).
  *
  * @param {Date|Number} date - the original date
+ * @param {Object} [options] - an object with options.
+ * @param {0|1|2|3|4|5|6|7|8|9|10|11} [options.yearStartsOn=0] - the index of the first month of the year (0 - January)
  * @returns {Date} the end of a year
  * @throws {TypeError} 1 argument required
  *
  * @example
  * // The end of a year for 2 September 2014 11:55:00:
- * var result = endOfYear(new Date(2014, 8, 2, 11, 55, 00))
+ * const result = endOfYear(new Date(2014, 8, 2, 11, 55, 00))
  * //=> Wed Dec 31 2014 23:59:59.999
+ *
+ * // The end of a year for 2 September 2014 11:55:00:
+ * const result = endOfYear(new Date(2014, 8, 2, 11, 55, 00), { yearStartsOn: 4 })
+ * //=> Thu Apr 30 2015 23:59:59.999
  */
-export default function endOfYear(dirtyDate: Date | number): Date {
+export default function endOfYear(dirtyDate: Date | number, options?: YearStart): Date {
   requiredArgs(1, arguments)
 
-  var date = toDate(dirtyDate)
-  var year = date.getFullYear()
-  date.setFullYear(year + 1, 0, 0)
-  date.setHours(23, 59, 59, 999)
+  let date = toDate(dirtyDate)
+  const year = date.getFullYear()
+
+  if (options) {
```

Comment:
Same as above, if options change, this will break.

```suggestion
  if (typeof options?.yearStartsOn === 'number') {
```